HDFS-15430. create should work when parent dir is internalDir and fallback configured.

umamaheswararao opened the pull request:
https://issues.apache.org/jira/browse/HDFS-15430

jojochuang commented on hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/fs/viewfs/TestViewFileSystemLinkFallback.java:
```diff
@@ -765,4 +766,154 @@ public void testMkdirsShouldReturnFalseWhenFallbackFSNotAvailable()
+    ConfigUtil.addLinkFallback(conf, fallbackTarget.toUri());
+    try (FileSystem vfs = FileSystem.get(viewFsDefaultClusterUri, conf)) {
+      Path p = new Path("/user2/test.file");
+      Path test = Path.mergePaths(fallbackTarget, p);
```
give this variable a more meaningful name?

umamaheswararao replied:
Done. Thanks

jojochuang commented on hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/fs/viewfs/TestViewFileSystemLinkFallback.java:
```diff
@@ -765,4 +766,154 @@ public void testMkdirsShouldReturnFalseWhenFallbackFSNotAvailable()
+    Path dir1 = new Path(targetTestRoot,
+        "fallbackDir/user1/hive/warehouse/partition-0");
+    fsTarget.mkdirs(dir1);
+    Path fallbackTarget = new Path(targetTestRoot, "fallbackDir");
```
Maybe easier to read if written as
`

    Path fallbackTarget = new Path(targetTestRoot, "fallbackDir");
    Path dir1 = new Path(fallbackTarget,
        "user1/hive/warehouse/partition-0");
    fsTarget.mkdirs(dir1);
`

umamaheswararao replied:
Done. Thanks

umamaheswararao commented on hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/fs/viewfs/TestViewFsLinkFallback.java:
```diff
@@ -294,4 +298,155 @@ public void testMkdirShouldFailWhenFallbackFSNotAvailable()
+  }
+
+  /**
+   * Tests the create of a file where he path is same as one of of the internal
```
Done.

umamaheswararao commented:
Thanks @jojochuang for the review. I have updated with fixing the comments. Let me know if you have further. Thanks

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  27m 28s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 2 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   1m 18s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  22m 36s |  trunk passed  |
| +1 :green_heart: |  compile  |  23m 12s |  trunk passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  compile  |  19m 21s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  checkstyle  |   3m 15s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |  21m 22s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  15m 59s |  branch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 31s |  root in trunk failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   5m 43s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +0 :ok: |  spotbugs  |  39m 12s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| -1 :x: |  findbugs  |  39m  7s |  root in trunk has 3 extant findbugs warnings.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 36s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |  26m  2s |  the patch passed  |
| +1 :green_heart: |  compile  |  21m 58s |  the patch passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  javac  |  21m 58s |  the patch passed  |
| +1 :green_heart: |  compile  |  19m 31s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  javac  |  19m 31s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   3m  0s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |  19m 14s |  the patch passed  |
| +1 :green_heart: |  shellcheck  |   0m  0s |  There were no new shellcheck issues.  |
| +1 :green_heart: |  shelldocs  |   0m 14s |  There were no new shelldocs issues.  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  15m 36s |  patch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 30s |  root in the patch failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   5m 32s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| -1 :x: |  findbugs  |   2m 12s |  hadoop-common-project/hadoop-common generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0)  |
| -1 :x: |  findbugs  |  32m 26s |  root generated 2 new + 3 unchanged - 0 fixed = 5 total (was 3)  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 590m  2s |  root in the patch passed.  |
| -1 :x: |  asflicense  |   1m 25s |  The patch generated 1 ASF License warnings.  |
|  |   | 931m 15s |   |


| Reason | Tests |
|-------:|:------|
| FindBugs | module:hadoop-common-project/hadoop-common |
|  |  Possible null pointer dereference of f in org.apache.hadoop.fs.viewfs.ViewFileSystem$InternalDirOfViewFs.create(Path, FsPermission, boolean, int, short, long, Progressable)  Dereferenced at ViewFileSystem.java:f in org.apache.hadoop.fs.viewfs.ViewFileSystem$InternalDirOfViewFs.create(Path, FsPermission, boolean, int, short, long, Progressable)  Dereferenced at ViewFileSystem.java:[line 1205] |
|  |  Possible null pointer dereference of f in org.apache.hadoop.fs.viewfs.ViewFs$InternalDirOfViewFs.createInternal(Path, EnumSet, FsPermission, int, short, long, Progressable, Options$ChecksumOpt, boolean)  Dereferenced at ViewFs.java:f in org.apache.hadoop.fs.viewfs.ViewFs$InternalDirOfViewFs.createInternal(Path, EnumSet, FsPermission, int, short, long, Progressable, Options$ChecksumOpt, boolean)  Dereferenced at ViewFs.java:[line 941] |
| FindBugs | module:root |
|  |  Possible null pointer dereference of f in org.apache.hadoop.fs.viewfs.ViewFileSystem$InternalDirOfViewFs.create(Path, FsPermission, boolean, int, short, long, Progressable)  Dereferenced at ViewFileSystem.java:f in org.apache.hadoop.fs.viewfs.ViewFileSystem$InternalDirOfViewFs.create(Path, FsPermission, boolean, int, short, long, Progressable)  Dereferenced at ViewFileSystem.java:[line 1205] |
|  |  Possible null pointer dereference of f in org.apache.hadoop.fs.viewfs.ViewFs$InternalDirOfViewFs.createInternal(Path, EnumSet, FsPermission, int, short, long, Progressable, Options$ChecksumOpt, boolean)  Dereferenced at ViewFs.java:f in org.apache.hadoop.fs.viewfs.ViewFs$InternalDirOfViewFs.createInternal(Path, EnumSet, FsPermission, int, short, long, Progressable, Options$ChecksumOpt, boolean)  Dereferenced at ViewFs.java:[line 941] |
| Failed junit tests | hadoop.yarn.applications.distributedshell.TestDistributedShell |
|   | hadoop.yarn.server.resourcemanager.scheduler.fair.TestFairScheduler |
|   | hadoop.yarn.server.resourcemanager.scheduler.fair.TestFairSchedulerPreemption |
|   | hadoop.yarn.server.resourcemanager.TestRMRestart |
|   | hadoop.yarn.server.resourcemanager.TestRMHATimelineCollectors |
|   | hadoop.hdfs.server.sps.TestExternalStoragePolicySatisfier |
|   | hadoop.fs.viewfs.TestViewFSOverloadSchemeWithMountTableConfigInHDFS |
|   | hadoop.hdfs.server.blockmanagement.TestBlockTokenWithDFSStriped |
|   | hadoop.hdfs.TestDFSOutputStream |
|   | hadoop.fs.viewfs.TestViewFileSystemOverloadSchemeWithHdfsScheme |
|   | hadoop.hdfs.server.datanode.TestBPOfferService |
|   | hadoop.hdfs.server.namenode.TestNameNodeRetryCacheMetrics |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2107 |
| Optional Tests | dupname asflicense shellcheck shelldocs compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux e0df4a858f71 4.15.0-101-generic #102-Ubuntu SMP Mon May 11 10:07:26 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / e8dc862d385 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/3/artifact/out/branch-javadoc-root-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/3/artifact/out/branch-findbugs-root-warnings.html |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/3/artifact/out/patch-javadoc-root-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/3/artifact/out/new-findbugs-hadoop-common-project_hadoop-common.html |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/3/artifact/out/new-findbugs-root.html |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/3/artifact/out/patch-unit-root.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/3/testReport/ |
| asflicense | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/3/artifact/out/patch-asflicense-problems.txt |
| Max. process+thread count | 3072 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs . U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/3/console |
| versions | git=2.17.1 maven=3.6.0 shellcheck=0.4.6 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   2m 31s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 2 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   1m 18s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  23m  1s |  trunk passed  |
| +1 :green_heart: |  compile  |  26m 23s |  trunk passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  compile  |  23m 13s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  checkstyle  |   4m 11s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |  23m 33s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  17m 48s |  branch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 41s |  root in trunk failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   7m 34s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +0 :ok: |  spotbugs  |  41m 33s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| -1 :x: |  findbugs  |  41m 24s |  root in trunk has 3 extant findbugs warnings.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 31s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |  30m 42s |  the patch passed  |
| +1 :green_heart: |  compile  |  26m 19s |  the patch passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  javac  |  26m 19s |  the patch passed  |
| +1 :green_heart: |  compile  |  22m 51s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  javac  |  22m 51s |  the patch passed  |
| -0 :warning: |  checkstyle  |   3m 48s |  root: The patch generated 4 new + 90 unchanged - 0 fixed = 94 total (was 90)  |
| +1 :green_heart: |  mvnsite  |  26m 11s |  the patch passed  |
| +1 :green_heart: |  shellcheck  |   0m  0s |  There were no new shellcheck issues.  |
| +1 :green_heart: |  shelldocs  |   0m 16s |  There were no new shelldocs issues.  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  15m 47s |  patch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 33s |  root in the patch failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   6m 29s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  findbugs  |  41m 57s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 644m  3s |  root in the patch passed.  |
| -1 :x: |  asflicense  |   1m 56s |  The patch generated 1 ASF License warnings.  |
|  |   | 1003m 20s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.yarn.server.nodemanager.webapp.TestNMWebServices |
|   | hadoop.yarn.server.resourcemanager.scheduler.fair.TestFairSchedulerPreemption |
|   | hadoop.yarn.server.resourcemanager.scheduler.fair.TestFairSchedulerOvercommit |
|   | hadoop.yarn.server.resourcemanager.TestRMHATimelineCollectors |
|   | hadoop.yarn.applications.distributedshell.TestDistributedShell |
|   | hadoop.fs.viewfs.TestViewFileSystemLinkFallback |
|   | hadoop.hdfs.server.namenode.TestCheckpoint |
|   | hadoop.hdfs.server.namenode.ha.TestBootstrapStandby |
|   | hadoop.fs.viewfs.TestViewFSOverloadSchemeWithMountTableConfigInHDFS |
|   | hadoop.fs.viewfs.TestViewFileSystemOverloadSchemeWithHdfsScheme |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2107 |
| Optional Tests | dupname asflicense shellcheck shelldocs compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux e1f2917693ed 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / e8dc862d385 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/2/artifact/out/branch-javadoc-root-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/2/artifact/out/branch-findbugs-root-warnings.html |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/2/artifact/out/diff-checkstyle-root.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/2/artifact/out/patch-javadoc-root-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/2/artifact/out/patch-unit-root.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/2/testReport/ |
| asflicense | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/2/artifact/out/patch-asflicense-problems.txt |
| Max. process+thread count | 2935 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs . U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/2/console |
| versions | git=2.17.1 maven=3.6.0 shellcheck=0.4.6 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 26s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 2 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   1m 17s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  25m 20s |  trunk passed  |
| +1 :green_heart: |  compile  |  24m 55s |  trunk passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  compile  |  21m 15s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  checkstyle  |   3m 19s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |  21m 39s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  16m 58s |  branch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 31s |  root in trunk failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   6m 45s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +0 :ok: |  spotbugs  |  38m 57s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| -1 :x: |  findbugs  |  38m 50s |  root in trunk has 3 extant findbugs warnings.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 30s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |  26m 37s |  the patch passed  |
| +1 :green_heart: |  compile  |  24m  8s |  the patch passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  javac  |  24m  8s |  the patch passed  |
| +1 :green_heart: |  compile  |  20m 48s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  javac  |  20m 48s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   3m 21s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |  21m 31s |  the patch passed  |
| +1 :green_heart: |  shellcheck  |   0m  0s |  There were no new shellcheck issues.  |
| +1 :green_heart: |  shelldocs  |   0m 14s |  There were no new shelldocs issues.  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  16m  6s |  patch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 31s |  root in the patch failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   6m 36s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| -1 :x: |  findbugs  |   2m 34s |  hadoop-common-project/hadoop-common generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0)  |
| -1 :x: |  findbugs  |  40m 52s |  root generated 2 new + 3 unchanged - 0 fixed = 5 total (was 3)  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 617m 47s |  root in the patch passed.  |
| -1 :x: |  asflicense  |   1m 24s |  The patch generated 1 ASF License warnings.  |
|  |   | 959m 18s |   |


| Reason | Tests |
|-------:|:------|
| FindBugs | module:hadoop-common-project/hadoop-common |
|  |  Possible null pointer dereference of f in org.apache.hadoop.fs.viewfs.ViewFileSystem$InternalDirOfViewFs.create(Path, FsPermission, boolean, int, short, long, Progressable)  Dereferenced at ViewFileSystem.java:f in org.apache.hadoop.fs.viewfs.ViewFileSystem$InternalDirOfViewFs.create(Path, FsPermission, boolean, int, short, long, Progressable)  Dereferenced at ViewFileSystem.java:[line 1205] |
|  |  Possible null pointer dereference of f in org.apache.hadoop.fs.viewfs.ViewFs$InternalDirOfViewFs.createInternal(Path, EnumSet, FsPermission, int, short, long, Progressable, Options$ChecksumOpt, boolean)  Dereferenced at ViewFs.java:f in org.apache.hadoop.fs.viewfs.ViewFs$InternalDirOfViewFs.createInternal(Path, EnumSet, FsPermission, int, short, long, Progressable, Options$ChecksumOpt, boolean)  Dereferenced at ViewFs.java:[line 941] |
| FindBugs | module:root |
|  |  Possible null pointer dereference of f in org.apache.hadoop.fs.viewfs.ViewFileSystem$InternalDirOfViewFs.create(Path, FsPermission, boolean, int, short, long, Progressable)  Dereferenced at ViewFileSystem.java:f in org.apache.hadoop.fs.viewfs.ViewFileSystem$InternalDirOfViewFs.create(Path, FsPermission, boolean, int, short, long, Progressable)  Dereferenced at ViewFileSystem.java:[line 1205] |
|  |  Possible null pointer dereference of f in org.apache.hadoop.fs.viewfs.ViewFs$InternalDirOfViewFs.createInternal(Path, EnumSet, FsPermission, int, short, long, Progressable, Options$ChecksumOpt, boolean)  Dereferenced at ViewFs.java:f in org.apache.hadoop.fs.viewfs.ViewFs$InternalDirOfViewFs.createInternal(Path, EnumSet, FsPermission, int, short, long, Progressable, Options$ChecksumOpt, boolean)  Dereferenced at ViewFs.java:[line 941] |
| Failed junit tests | hadoop.hdfs.server.datanode.TestBPOfferService |
|   | hadoop.fs.viewfs.TestViewFileSystemOverloadSchemeWithHdfsScheme |
|   | hadoop.fs.viewfs.TestViewFSOverloadSchemeWithMountTableConfigInHDFS |
|   | hadoop.hdfs.server.balancer.TestBalancer |
|   | hadoop.hdfs.server.namenode.ha.TestHAAppend |
|   | hadoop.yarn.sls.appmaster.TestAMSimulator |
|   | hadoop.yarn.applications.distributedshell.TestDistributedShell |
|   | hadoop.yarn.server.resourcemanager.TestRMHATimelineCollectors |
|   | hadoop.yarn.server.resourcemanager.scheduler.fair.TestFairSchedulerPreemption |
|   | hadoop.metrics2.source.TestJvmMetrics |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2107 |
| Optional Tests | dupname asflicense shellcheck shelldocs compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux f11151fd5d90 4.15.0-101-generic #102-Ubuntu SMP Mon May 11 10:07:26 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / e8dc862d385 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/4/artifact/out/branch-javadoc-root-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/4/artifact/out/branch-findbugs-root-warnings.html |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/4/artifact/out/patch-javadoc-root-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/4/artifact/out/new-findbugs-hadoop-common-project_hadoop-common.html |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/4/artifact/out/new-findbugs-root.html |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/4/artifact/out/patch-unit-root.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/4/testReport/ |
| asflicense | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/4/artifact/out/patch-asflicense-problems.txt |
| Max. process+thread count | 3054 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs . U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/4/console |
| versions | git=2.17.1 maven=3.6.0 shellcheck=0.4.6 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m  0s |  Docker mode activated.  |
| -1 :x: |  patch  |   0m  4s |  https://github.com/apache/hadoop/pull/2107 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help.  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/2107 |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/5/console |
| versions | git=2.17.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 11s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 3 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 22s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  21m 20s |  trunk passed  |
| +1 :green_heart: |  compile  |  20m 39s |  trunk passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  compile  |  17m 24s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  checkstyle  |   2m 49s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   2m 43s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  21m 32s |  branch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 36s |  hadoop-common in trunk failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| -1 :x: |  javadoc  |   0m 41s |  hadoop-hdfs in trunk failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   1m 43s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +0 :ok: |  spotbugs  |   3m 14s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   5m 19s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 23s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m  1s |  the patch passed  |
| +1 :green_heart: |  compile  |  19m 55s |  the patch passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  javac  |  19m 55s |  the patch passed  |
| +1 :green_heart: |  compile  |  17m 22s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  javac  |  17m 22s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   2m 52s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   2m 44s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  15m 45s |  patch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 35s |  hadoop-common in the patch failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| -1 :x: |  javadoc  |   0m 42s |  hadoop-hdfs in the patch failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   1m 43s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  findbugs  |   5m 32s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   9m 31s |  hadoop-common in the patch passed.  |
| -1 :x: |  unit  | 112m 26s |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 54s |  The patch does not generate ASF License warnings.  |
|  |   | 287m 37s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.TestBlocksScheduledCounter |
|   | hadoop.hdfs.server.namenode.TestNameNodeRetryCacheMetrics |
|   | hadoop.hdfs.TestBlockStoragePolicy |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2107 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 2cd5443e0f3c 4.15.0-101-generic #102-Ubuntu SMP Mon May 11 10:07:26 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 04abd0eb17b |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/6/artifact/out/branch-javadoc-hadoop-common-project_hadoop-common-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/6/artifact/out/branch-javadoc-hadoop-hdfs-project_hadoop-hdfs-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/6/artifact/out/patch-javadoc-hadoop-common-project_hadoop-common-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/6/artifact/out/patch-javadoc-hadoop-hdfs-project_hadoop-hdfs-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/6/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/6/testReport/ |
| Max. process+thread count | 3238 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/6/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

umamaheswararao commented:
Test failures are unrelated and javadoc failures due to HADOOP-17091.

ayushtkn commented on hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/fs/viewfs/TestViewFileSystemLinkFallback.java:
```diff
@@ -765,4 +766,151 @@ public void testMkdirsShouldReturnFalseWhenFallbackFSNotAvailable()
+        vfs.create(vfsTestDir);
+        Assert.fail("Should fail to create file as this is an internal dir.");
+      } catch (NotInMountpointException e){
+        // This tree is part of internal tree. The above expetion will be thrown
```
typo "expetion"

umamaheswararao replied:
Thanks. Done.

ayushtkn commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFs.java:
```diff
@@ -919,6 +921,42 @@ public FSDataOutputStream createInternal(final Path f,
         ParentNotDirectoryException, UnsupportedFileSystemException,
         UnresolvedLinkException, IOException {
+      Preconditions.checkNotNull(f, "File cannot be null.");
+      // Just a sanity check. This should not happen.
```
What do you mean by : This should not happen. Ideally client call shouldn't land here or This If check will never get hit(It does get hit, I verified),

umamaheswararao replied:
My comments seems to be too light. "//" will trigger this. But ideally that should not happen as users would not create files with /. However if somebody attempts we are ready to shoot. I have removed that comment, as it may be confusing.
However, this will trigger mostly in ViewFs.java, but may not be in ViewFileSystem.java. In ViewFileSystem.java, getDefaultReplication API itself might shade this condition if parent is internal currently. Anyway the current condition should be good enough to handle if some one try with files as "//" etc
Thanks for checking.

umamaheswararao commented:
Thank you @ayushtkn for review!!
I have just corrected typo and removed the confusing comment. Since these are just comment correction/removal, I will proceed to commit. Thank you.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 11s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 3 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   1m  2s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  21m 37s |  trunk passed  |
| +1 :green_heart: |  compile  |  21m 25s |  trunk passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  compile  |  17m 39s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  checkstyle  |   2m 49s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   2m 45s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  21m 32s |  branch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 37s |  hadoop-common in trunk failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| -1 :x: |  javadoc  |   0m 41s |  hadoop-hdfs in trunk failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   1m 44s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +0 :ok: |  spotbugs  |   3m 13s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   5m 17s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 22s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m  0s |  the patch passed  |
| +1 :green_heart: |  compile  |  20m 23s |  the patch passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  javac  |  20m 23s |  the patch passed  |
| +1 :green_heart: |  compile  |  17m 41s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  javac  |  17m 41s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   2m 52s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   2m 47s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  15m 34s |  patch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 36s |  hadoop-common in the patch failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| -1 :x: |  javadoc  |   0m 42s |  hadoop-hdfs in the patch failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   1m 45s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  findbugs  |   5m 36s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   9m 32s |  hadoop-common in the patch passed.  |
| -1 :x: |  unit  | 121m 50s |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 58s |  The patch does not generate ASF License warnings.  |
|  |   | 300m  5s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.sps.TestExternalStoragePolicySatisfier |
|   | hadoop.hdfs.TestDFSOutputStream |
|   | hadoop.hdfs.server.namenode.TestNameNodeRetryCacheMetrics |
|   | hadoop.hdfs.TestStripedFileAppend |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/7/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2107 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 2f7064d44883 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / e0cededfbd2 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/7/artifact/out/branch-javadoc-hadoop-common-project_hadoop-common-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/7/artifact/out/branch-javadoc-hadoop-hdfs-project_hadoop-hdfs-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/7/artifact/out/patch-javadoc-hadoop-common-project_hadoop-common-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/7/artifact/out/patch-javadoc-hadoop-hdfs-project_hadoop-hdfs-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/7/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/7/testReport/ |
| Max. process+thread count | 2722 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2107/7/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.
